src/components/gltf-model-plus.js

```diff
@@ -11,6 +11,8 @@ import { RGBELoader } from "three/examples/jsm/loaders/RGBELoader";
 import { BasisTextureLoader } from "three/examples/jsm/loaders/BasisTextureLoader";
 import { DRACOLoader } from "three/examples/jsm/loaders/DRACOLoader";
 import { GLTFLoader } from "three/examples/jsm/loaders/GLTFLoader";
+import GLBRangeRequests from "three-gltf-extensions/loaders/GLB_range_requests/GLB_range_requests";
+import GLTFLodExtension from "three-gltf-extensions/loaders/MSFT_lod/MSFT_lod";
```

Comment:
No good way to comment upstream since the there are multiple closed PRs so just including links here:

https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L224
Based on your comments upstream I agree that you should be able to handle the material LOD case with just node LODs (making other node LOD levels just reference a different material), so shouldn't we prefer node rather than material in the case both exist?

https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L291
Shouldn't `onLoadMesh`` be called here as well like we do above for the material LOD case?

https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L120
This LOD swapping with dummy objects in onBeforeRender feels slightly hacky but seems like it should work. I mostly don't like that its happening at a sort of unpredictable part in the frame since its both in `onBeforeRender` but also a `.then`. I wonder if we can think through a way to do this as system instead. Maybe after we have transitioned gltf loading to bitecs fully we can revisit.

https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L115
I assume (though I don't actually see it specifically outlined in the spec) that "coverage" is a 0-1 % of the screen that rendering the object will cover when this LOD should be used. It's not really clear how you would translate that to distance as its actually going to change at runtime depending on the camera properties. In practice just using a distance seems fine for Hubs but I don't think it would be to spec. Like the previous one maybe this is something we would handle in an "LOD system".

https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L12
After this `getCurrentLevel` could be pointing at the level that was just removed. If it was one of the middle levels I guess it would still be accurate since it would just be pointing at the next level but if the removed level was the last it would be pointing at an invalid index.

https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L159
Hmm yeah I think the issue is coming into play here because THREE.LOD is at the node level. We almost want something like an LODMaterial that would do similar swapping of materials based on distanced. Though honeslty I am a bit confused what the usecase for material LODs is. Seems like any case could be handled by node LODs.

Reply:
And some general notes:
When I was originally thinking about LODs I assumed that the different LOD levels would only be pointing at different meshes (like what was proposed here https://github.com/KhronosGroup/glTF/issues/1045#issuecomment-457059262), so only the Geometry in THREE would need switching. It seems MSFT_LOD switches at the Node level, which means technically different LODs could have different sets of Hubs components. We would need to account for that in the "progressive" and "any" case since not all the component data would be in the scene graph when its first processed in gtlf-model-plus. Maybe we just say this is unsupported for now. If it seems useful we can implement it as part of rewriting gltf loading as we move away from aframe rather than making this more complex only to get rid of it later.

Related. It might end up being a bit of an issue for some components that they will now be "on" the THREE.LOD instead of a THREE.Mesh. (well until we fully migrate away from aframe they are "on" the THREE.Group which aframe creates, but .getObject3D("mesh") will now be a THREE.LOD)

Reply:
Also just realized the additional layer of the **THREE.LOD** node may cause issues with animations, but maybe THREE already accounts for this.

Reply:
> https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L224
> Based on your comments upstream I agree that you should be able to handle the material LOD case with just node LODs (making other node LOD levels just reference a different material), so shouldn't we prefer node rather than material in the case both exist?

The reason why I give the priority material LOD over node LOD is just the implementation can be simpler due to `GLTFLoader` plugin API and design. No other big reasons.

And the behavior of the case where the both node and material LODs are defined is not clarified in the spec. Personally I think the spec shouldn't allow the both LOD definitions and/or the spec may drop material LOD support for simplicity as suggested in

https://github.com/KhronosGroup/glTF/issues/1952

I'm not sure if we really want to take effort for the case where both node and material LODs are defined because I'm not sure if there are many assets having the both. So, I want to go as is for now and think more later if needed.

Reply:
> https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L291
Shouldn't `onLoadMesh`` be called here as well like we do above for the material LOD case?

Good catch, I updated the plugin. I will upgrade the package when I will resolve all the issues.

Reply:
> https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L120
This LOD swapping with dummy objects in onBeforeRender feels slightly hacky but seems like it should work. I mostly don't like that its happening at a sort of unpredictable part in the frame since its both in onBeforeRender but also a .then. I wonder if we can think through a way to do this as system instead. Maybe after we have transitioned gltf loading to bitecs fully we can revisit.

Honestly, I don't really like this hack using `onBeforeRender` neither.

The `three-gltf-extensions` `MSFT_lod` plugin is written for generic use and is designed as no or minimul user code requirement.

Yes, we may rewrite it as system after we finish the refactoring most of the core client code with bitecs.

Reply:
> https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L115
I assume (though I don't actually see it specifically outlined in the spec) that "coverage" is a 0-1 % of the screen that rendering the object will cover when this LOD should be used. It's not really clear how you would translate that to distance as its actually going to change at runtime depending on the camera properties. In practice just using a distance seems fine for Hubs but I don't think it would be to spec. Like the previous one maybe this is something we would handle in an "LOD system".

Yes, `MSFT_screencoverage` is (probably) screen coverage ratio as the word indicates.

But Hubs allows to scale objects. If we take scale into account, we may need to observe the scale change and update the distance when it is updated.

And Three.js uses distance from camera, not screen coverage, to choose a LOD level. Accurately simulating LOD level switch based on screen coverage with `THREE.LOD` is hard. For example imagine that a mesh is at corner of the viewport and half of the body is out of the screen.

And to calculate the screen coverage camera parameters info is needed. But I didn't want the plugin to have the dependency with camera (unless it's really needed).

Accuratly simulating the LOD level switch based on screen coverage in Three.js may be possible but can make the code complex. And honestly, I'm not really sure how much worth to accurately simulating it.

So I decided to use a simple and rough conversion for now. We can revisit later if needed.

Reply:
> https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L159
Hmm yeah I think the issue is coming into play here because THREE.LOD is at the node level. We almost want something like an LODMaterial that would do similar swapping of materials based on distanced. Though honeslty I am a bit confused what the usecase for material LODs is. Seems like any case could be handled by node LODs.

Because of the current `GLTFLoader` plugin API and design, I thought the solution polluting json may be the easiest way and other solutions can be much more complex.

> Though honeslty I am a bit confused what the usecase for material LODs is.

Same here. Honestly I think it may be ok not to support material LOD. (And also I suggested to remove material LOD from `MSFT_lod`.)

I didn't want to take effort for LOD material so that I used that possibly bad solution.

Reply:
> https://github.com/takahirox/three-gltf-extensions/blob/62288914fbafb2500e500b0209b33d4f3a96cfac/loaders/MSFT_lod/MSFT_lod.js#L12
After this getCurrentLevel could be pointing at the level that was just removed. If it was one of the middle levels I guess it would still be accurate since it would just be pointing at the next level but if the removed level was the last it would be pointing at an invalid index.

`removeLevel()` is called only right before `LOD.addLevel()` call and `getCurrentLevel` is re-calculated in the renderer in the next frame. So I don't think it is a big deal.

Reply:
> When I was originally thinking about LODs I assumed that the different LOD levels would only be pointing at different meshes (like what was proposed here https://github.com/KhronosGroup/glTF/issues/1045#issuecomment-457059262), so only the Geometry in THREE would need switching.

Honestly I agree with that comment. I think LOD extension should allow only LOD meshes in a node like

```
node: [
  {
    mesh: 0,
    extensions: {
      EXT_lod: {
        indices: [ // mesh indices
          1,
          2
        ]
    }
  }
]
```

The current asset with `MSFT_lod` spec and its handler can be more complex than it really needs to be.

I'm thinking to suggest it to simplify `MSFT_lod`, or propose a new LOD extension like `EXT_mesh_lod`.

By the way, `Mesh` is a set of geometry and material. Even if the LOD extension allows only mesh LODs in a node, switching only Geometry in Three.js isn't good enough, switching Mesh is needed.

Reply:
> Also just realized the additional layer of the THREE.LOD node may cause issues with animations, but maybe THREE already accounts for this.

Mesh objects are placed under LOD and key frame animation is bound to LOD object so keyframe position/rotation/scale animation shouldn't be a problem because they are propagated to child objects world matrices.

Reply:
> By the way, Mesh is a set of geometry and material. Even if the LOD extension allows only mesh LODs in a node, switching only Geometry in Three.js isn't good enough, switching Mesh is needed.

Ah right. For some reason I thought material was on `node` in GLTF but its not. We will definitely have to figure out what to do about components. I guess in the new world this isn't so bad since components are associated with an eid, and an eid is also associated with an Object3D, but you should not hold onto a reference to that Object3D ever, so it should be fine for us to swap it for LODs. We actually do this here https://github.com/mozilla/hubs/blob/38d5cbfd21a3bf0ce8d3ddc01132e5c702f9b72b/src/bit-systems/video-system.ts#L79 though we are still a bit unsure if its a good idea.

> Mesh objects are placed under LOD and key frame animation clip is bound to LOD object so keyframe position/rotation/scale animation shouldn't be a problem because they are propagated to child objects world matrices.

Right but if an animation targets a child of something that becomes an Three.LOD won't it be an issue because there is now an extra parent  `LOD -> Obj A -> Obj B`. When we added the THREE.Group nodes for AFRAME i think we also had to rewrite animations to target.